react-js/react-part-2.md

```diff
@@ -0,0 +1,219 @@
+# React Part 2
+
+## Projected Time
+
+2 hours 30 mins
+
+- Lesson: 1 hour
+- Guided Practice: 30 mins
+- Independent Practice: 1 hour
+
+### Prerequisites
+
+- [React](react.md)
+
+### Motivation
+
+In introduction to [React](/react.md) we discussed about how React makes websites interactive and reduces jQuery code. Now the problem arises if we want to store some information which intent to change and due to change of that information the UI changes i.e the page is re-rendered. Thus, component state is used for such cases.
+
+### Objectives
+
+- Component state
+- State vs Props
+- State are Immutable
+
+### Lesson
+
+- [Introduction to component state in React (video)](https://www.youtube.com/watch?v=e5n9j9n83OM)
+- [Component state in React slides](https://docs.google.com/presentation/d/1Trj_7vWYe79_mDboDxN9rdT94yjY5Dt55xtU1k3dUsw/edit?usp=sharing)
+
+### Common Mistakes & Misconceptions
+
+1. The scope of the state is local to the component.
+2. Don't forget to update state immutably.
+
+### Guided Practice
+
+- Define state.
+  ```
+  state = {}
+  ```
+- Initialize state
+
+  - State in class component can be initialized in two ways
+
+    - Directly inside the class component, outside all methods.
+
+      ```
+      state = {
+        message: "Hello world"
+      }
+      ```
+
+    - Inside class constructor
+      ```
+      constructor(){
+        this.state = {
+          message: "Hello world"
+          }
+      }
+      ```
+
+- State should not be mutated directly i.e. a copy of state should be created and then updated.
+  - Incorrect way -
+    ```
+    this.state.list.push(value)
+    ```
+  - Correct way -
+    ```
+        this.setState({
+          ...this.state,
+          list: [ ...this.state.list, value ]
+        })
+    ```
+- You must have understand till now that we should not update state directly, but what special thing _setState_ function does to the state, that the component re-renders automatically. Well, in introduction to React, we have learnt that React follows a special path called the React cycle. So, when state is updated using setState, React goes into the **React cycle** and then the changes are compared from the previous **React virtual DOM element**, changes are done accordingly and those changes are reflected in the UI.
+
+#### Now, let us create a simple todo list.
+
+1. First of all get rid of all the unnecessary stuff showing on DOM, remove all the lines in render of `App.js`.
+
+   ```
+    import React from 'react';
```

Comment:
Can you create a starting CodePen for this with the right libraries and syntax? For example, there will be no `import` or `export`.

here is a tiny example: https://codepen.io/vegetabill/pen/rNOMZGK